### PR TITLE
Graph Selection in OSSMC

### LINKED ISSUE #173: Graph Selection in OSSMC plugin
Now that PFT graph has been added to Kiali code, it is time to be able to select between Cytoscape and PFT graph in the OSSMC plugin. PFT graph should be the default option.

### PR DESCRIPTION
**Pull request by ferhoyos**
Closes #173.

Testing PR:
Graph selection can be set in OSSMConsole CR (`spec.kiali.graph`) field. It overwrites any option selected in Kiali CR (`spec.kiali_feature_flags.ui_defaults.graph.impl`). In development the graph value can be modified in plugin-config.json file (see README).

Unlike Kiali, here it is not possible to have both graphs at the same time ('both' is not a valid option). I have not found any way to create another graph section in a conditional way.

Some snapshots of Patternfly graph in OSSMC.

- Graph section:
image

- Service detail
image

- Workload detail
image

- Project detail
image

### CODE REVIEW

File: operator/roles/default/ossmconsole-deploy/defaults/main.yml
Comment (jmazzitelli) [inline operator/roles/default/ossmconsole-deploy/defaults/main.yml:11]:
```suggestion
  version: "default"
```

File: operator/roles/default/ossmconsole-deploy/defaults/main.yml
Comment (jmazzitelli) [inline operator/roles/default/ossmconsole-deploy/defaults/main.yml:16]:
```suggestion
    imagePullPolicy: "IfNotPresent"
```

File: operator/roles/default/ossmconsole-deploy/defaults/main.yml
Comment (jmazzitelli) [inline operator/roles/default/ossmconsole-deploy/defaults/main.yml:25]:
```suggestion
    graph: "pf"
```

**Review by jmazzitelli — CHANGES_REQUESTED** [review]
This is trivial.... For the operator ansible yaml (especially in the defaults file here), I like to keep the double quotes to keep things consistent and to ensure people reading this file understand these must be strings. Yes, it should be obvious for some of these, but explicitly putting double quotes catches your eye, and it makes the number/integer fields stand out more clearly so you can easily pick out which must be number fields and which are strings. (note, if you see a string field NOT wrapped in quotes, go ahead and wrap them).

I think I'll have another trivial review coming :) But I'm still going through this.

File: operator/manifests/template/manifests/ossmconsole.clusterserviceversion.yaml
Comment (jmazzitelli) [inline operator/manifests/template/manifests/ossmconsole.clusterserviceversion.yaml]: revert:
```suggestion
```

File: operator/manifests/ossmconsole-ossm/manifests/ossmconsole.clusterserviceversion.yaml
Comment (jmazzitelli) [inline operator/manifests/ossmconsole-ossm/manifests/ossmconsole.clusterserviceversion.yaml]: revert:
```suggestion
              "servicePort": 0
```

File: operator/manifests/template/manifests/ossmconsole.clusterserviceversion.yaml
Comment (jmazzitelli) [inline operator/manifests/template/manifests/ossmconsole.clusterserviceversion.yaml]: revert:
```suggestion
              "servicePort": 0
```

File: operator/roles/default/ossmconsole-deploy/defaults/main.yml
Comment (ferhoyos) [reply operator/roles/default/ossmconsole-deploy/defaults/main.yml]: Done

**Comment by ferhoyos** [comment]
Thanks @jmazzitelli. Operator changes are done.

File: operator/crd-docs/cr/kiali.io_v1alpha1_ossmconsole.yaml
Comment (jmazzitelli) [inline operator/crd-docs/cr/kiali.io_v1alpha1_ossmconsole.yaml]: @jshaughn  re: the graph customizations in the plugin... Do you think we need to prepare for a dict of settings here for the graph (in order to support multiple settings in the future)?

Because right now, this means the `kiali.graph` OSSMC CR setting can only ever be a string (so it only supports this one value, and no other graph customization features can be added under the `graph` yaml here).

Do you think we might have additional graph settings that the OSSMC plugin will need?

In short, I wonder if we need to do this - thus leaving open the ability to add additional values in the future should we need them (plus, naming it `impl` makes the meaning of the setting more clear):

```suggestion
    graph:
      impl: "pf"
```

File: operator/roles/default/ossmconsole-deploy/templates/openshift/configmap-plugin.yaml
Comment (jmazzitelli) [inline operator/roles/default/ossmconsole-deploy/templates/openshift/configmap-plugin.yaml:10]: you want ossmconsole_vars.kiali.graph here.. something like:
```suggestion
      "graph": "{{ ossmconsole_vars.kiali.graph }}"
```

Comment (ferhoyos) [reply operator/roles/default/ossmconsole-deploy/templates/openshift/configmap-plugin.yaml]: Done

**Review by jmazzitelli — CHANGES_REQUESTED** [review]
merge this PR-to-the-PR to get a molecule test for the new graph setting:
https://github.com/ferhoyos/openshift-servicemesh-plugin/pull/5

**Comment by ferhoyos** [comment]
@jmazzitelli thanks for the new graph setting molecule test PR.

**Review by jmazzitelli — APPROVED** [review]
lgtm, and the molecule test passes.

File: operator/crd-docs/cr/kiali.io_v1alpha1_ossmconsole.yaml
Comment (jshaughn) [reply operator/crd-docs/cr/kiali.io_v1alpha1_ossmconsole.yaml]: I prefer the `.impl`, it opens up the ability to have other customization and also is consistent with the Kiali CR.

**Review by jshaughn — APPROVED** [review]
In a comment I mentioned that I prefer `graph.impl` for easier extension and for consistency with Kiali CR, but I see that the it would cause a lot of changes in the PR.  So, I'll leave it up to @ferhoyos.  I'll approve as-is, but if you choose to make the update I can re-approve.  Either way, nice work!

**Comment by ferhoyos** [comment]
> In a comment I mentioned that I prefer `graph.impl` for easier extension and for consistency with Kiali CR, but I see that the it would cause a lot of changes in the PR. So, I'll leave it up to @ferhoyos. I'll approve as-is, but if you choose to make the update I can re-approve. Either way, nice work!

Not a problem to rename the variable to `graph.impl` for easier extension and consistency with Kiali CR.

**Review by jmazzitelli — APPROVED** [review]
I ran it, looks good. I also validated the CR and the CRD ... all good.

```
./validate-ossmconsole-cr.sh -cn ossmconsole -n ossmconsole
=== SETTINGS ===
OSSMCONSOLE_CRD_LOCATION=
OSSMCONSOLE_CR_FILE=
OSSMCONSOLE_CR_NAME=ossmconsole
NAMESPACE=ossmconsole
PRINT_CRD=false
=== SETTINGS ===
Using 'oc' located here: /home/jmazzite/bin/oc
...
Validating the CR:
----------
testossmconsole.kiali.io/ossmconsole created
----------
OSSMConsole CR [ossmconsole] in namespace [ossmconsole] is valid.
```

(we should probably figure out how to add that validation to the CI workflow, but that's for another PR - that is done here: https://github.com/kiali/openshift-servicemesh-plugin/pull/178)

**Review by jshaughn — APPROVED** [review]
pf = perfect!

**Pull request merged**
